src/main/java/org/broadinstitute/hellbender/tools/walkers/featuremapping/AddFlowSNVQuality.java

```diff
@@ -0,0 +1,601 @@
+                        calledIndex = i;
+                    }
+                }
+
+                // qCalled
+                if ( calledIndex >= 0 ) {
+                    line.add(Integer.toString(qX[calledIndex][pos] - 33));
+                } else {
+                    line.add("-1");
+                }
+
+                // write the line
+                pw.println(StringUtils.join(line, ','));
+            }
+
+            // close file
+            pw.close();
+        } catch (IOException e) {
+            throw new GATKException("", e);
+        }
+    }
+
+    private void printStats(final String fname) throws IOException {
+
+        inputQualStats.csvWrite(fname + ".inputQual.csv");
+        outputBQStats.csvWrite(fname + ".outputBQ.csv");
+        outputQAltStats.csvWrite(fname + ".outputQAlt.csv");
+        outputQCalledStats.csvWrite(fname + ".outputQCalled.csv");
+        outputSumPStats.csvWrite(fname + ".outputSumP.csv");
+    }
+
+    static public String attrNameForNonCalledBase(byte nonCalledBase) {
+        return attrNameForNonCalledBase((char)nonCalledBase);
+    }
+
+    static public String attrNameForNonCalledBase(char nonCalledBase) {
+        return "q" + Character.toLowerCase(nonCalledBase);
+    }
+
+    public void addBaseQuality(final GATKRead read, final SAMFileHeader hdr, double limitPhoreScore, FlowBasedArgumentCollection fbargs) {
+
+        // take in phred score limit
+        if ( !Double.isNaN(limitPhoreScore) ) {
+            maxQualityScore = (int)limitPhoreScore;
+            minLikelihoodProbRate = Math.pow(10, -limitPhoreScore / 10.0);
+        }
+
+        // convert to a flow base read
+        final FlowBasedReadUtils.ReadGroupInfo rgInfo = FlowBasedReadUtils.getReadGroupInfo(hdr, read);
+        final FlowBasedRead fbRead = new FlowBasedRead(read, rgInfo.flowOrder, rgInfo.maxClass, fbargs);
+        final int flowOrderLength = calcFlowOrderLength(rgInfo.flowOrder);
+
+        // generate base quality
+        final AtomicReference<double[][]> snvResultRef = new AtomicReference<>();
+        final double[]      baseErrorProb = generateBaseErrorProbability(fbRead, flowOrderLength, rgInfo.flowOrder.getBytes(), snvResultRef);
+
+        // install in read
+        read.setAttribute(BASE_QUALITY_ATTRIBUTE_NAME, new String(convertErrorProbToPhred(baseErrorProb, true)));
+        for ( int i = 0 ; i < flowOrderLength ; i++ ) {
+            final String name = AddFlowSNVQuality.attrNameForNonCalledBase(rgInfo.flowOrder.charAt(i));
+            read.setAttribute(name, new String(convertErrorProbToPhred(snvResultRef.get()[i], true)));
+        }
+    }
+
+    private byte[] convertErrorProbToPhred(double[] errorProb, boolean asciiBase) {
+
+        final byte floor = asciiBase ? (byte)PHRED_ASCII_BASE : 0;
+
+        final byte[] phred = new byte[errorProb.length];
+        for ( int i = 0 ; i < errorProb.length ; i++ ) {
+            if ( errorProb[i] == 0 ) {
```

Comment:
Needs a comment I think: how come errorProb can be zero

Reply:
Comment added.

// normally an error probability can not be zero.
// still, this method is called to convert base quality as well as snvq, which is computed.
// the following check is a safety, in case snvq produces a zero.